src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -0,0 +1,183 @@
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: Bradley Wang
+ */
+
+#include "cpu/flexcpu/simple_dataflow_thread.hh"
+
+#include "mem/request.hh"
+
+using namespace std;
+
+// Fullsystem mode constructor
+SDCPUThread::SDCPUThread(SimpleDataflowCPU* cpu_, ThreadID tid_,
+                    System* system_, BaseTLB* itb_, BaseTLB* dtb_,
+                    TheISA::ISA* isa_, bool use_kernel_stats_ = true):
+    _cpuPtr(cpu_),
+    _committedState(new SimpleThread(cpu_, tid_, system_, itb_, dtb_, isa_,
+                                     use_kernel_stats_))
+{
+    //
+}
+
+// Non-fullsystem constructor
+SDCPUThread::SDCPUThread(SimpleDataflowCPU* cpu_, ThreadID tid_,
+                    System* system_, Process* process_, BaseTLB* itb_,
+                    BaseTLB* dtb_, TheISA::ISA* isa_):
+    _cpuPtr(cpu_),
+    _committedState(new SimpleThread(cpu_, tid_, system_, process_, itb_, dtb_,
+                                     isa_))
+{
+    //
+}
+
+SDCPUThread::~SDCPUThread()
+{
+    // TODO Properly hold state since THIS IS SUPER UNSAFE AND WILL RESULT IN
+    // MULTIPLE DELETION UPON COPY.
+    delete _committedState;
+}
+
+ThreadContext*
+SDCPUThread::getThreadContext()
+{
+    // TODO wrap the TC so we can capture accesses which should kill any
+    // speculative state.
+    return _committedState->getTC();
+}
+
+void
+SDCPUThread::issueInstruction(StaticInstPtr inst, TheISA::PCState cur_pc)
+{
+    const InstSeqNum seq_num = lastCommittedInstNum + inflightInsts.size();
+
+    shared_ptr<InflightInst> dynamic_inst =
+        make_shared<InflightInst>(getThreadContext(), seq_num, inst, cur_pc);
+
+    dynamic_inst->populateDataDependencies(inflightInsts, lastUses);
+    // TODO update lastUses to include the new instruction's uses
```

Comment:
again, I think this would be easier if `lastUses` was it's own type (e.g., dependencyGraph)